Enhance Association Request Timeout (1284)

BobSter3000 opened the pull request:
Update associatonRequestTimesout state in  DicomClientAssociationState

Fixes #1284.

#### Checklist
- [X] The pull request branch is in sync with latest commit on the *fo-dicom/development* branch
- [x] I have updated API documentation
- [x] I have included unit tests
- [X] I have updated the change log
- [X] I am listed in the CONTRIBUTORS file

#### Changes proposed in this pull request:
- Updates DicomClient so that the state transitions to TransitionToCompletedWithErrorState instead of TransitionToCompletedState when the association times out a few times (configurable via setting).
- Add new Exception for TransitionToCompletedWithErrorState

BobSter3000 committed:
Add Association Request Exception
Update associatonRequestTimesout state in  DicomClientAssociationState

amoerie commented:
Looks like a unit test is failing now, in the v5-net-framework build.
Since the net-core-21 and net-core-31 builds are succeeding, we can assume this bug doesn't always occur but is circumstantial.

This is the exception:

```
[xUnit.net 00:00:16.80]     FellowOakDicom.Tests.Network.AsyncDicomCEchoProviderTests.OnCEchoRequestAsync_ShouldRespond [FAIL]
  Failed FellowOakDicom.Tests.Network.AsyncDicomCEchoProviderTests.OnCEchoRequestAsync_ShouldRespond [12 s]
  Error Message:
   FellowOakDicom.Network.DicomAssociationRequestTimedoutException : Association Request timed out waiting for response
  Stack Trace:
     at FellowOakDicom.Network.Client.States.DicomClientCompletedState.<GetNextStateAsync>d__8.MoveNext() in D:\a\fo-dicom\fo-dicom\FO-DICOM.Core\Network\Client\States\DicomClientCompletedState.cs:line 127
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at FellowOakDicom.Network.Client.DicomClient.<Transition>d__91.MoveNext() in D:\a\fo-dicom\fo-dicom\FO-DICOM.Core\Network\Client\DicomClient.cs:line 267
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at FellowOakDicom.Network.Client.States.DicomClientConnectState.<GetNextStateAsync>d__5.MoveNext() in D:\a\fo-dicom\fo-dicom\FO-DICOM.Core\Network\Client\States\DicomClientConnectState.cs:line 82
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at FellowOakDicom.Network.Client.DicomClient.<Transition>d__91.MoveNext() in D:\a\fo-dicom\fo-dicom\FO-DICOM.Core\Network\Client\DicomClient.cs:line 267
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at FellowOakDicom.Network.Client.States.DicomClientIdleState.<SendAsync>d__5.MoveNext() in D:\a\fo-dicom\fo-dicom\FO-DICOM.Core\Network\Client\States\DicomClientIdleState.cs:line 58
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at FellowOakDicom.Network.Client.DicomClient.<SendAsync>d__110.MoveNext() in D:\a\fo-dicom\fo-dicom\FO-DICOM.Core\Network\Client\DicomClient.cs:line 377
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at FellowOakDicom.Tests.Network.AsyncDicomCEchoProviderTests.<OnCEchoRequestAsync_ShouldRespond>d__2.MoveNext() in D:\a\fo-dicom\fo-dicom\Tests\FO-DICOM.Tests\Network\AsyncDicomCEchoProviderTests.cs:line 51
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
```

There's a few things at play here:
- the default association request timeout is 5 seconds
- we are not guaranteed to have a very performant CI server
- even if we have a performant CI server, it can be busy with another test in parallel that is very demanding

Previously, if an association request times out, the DicomClient would cycle through its cleanup states and eventually retry.
You could argue this improves the stability of the DicomClient. 
On the other hand, if it keeps timing out, it's stuck in an endless loop, which is also not good.

I'm now starting to doubt whether this is an improvement. Wouldn't our efforts better be focused on adding an "escape hatch" if the association request keeps timing out? For example, we could expose a new event `DicomClient.AssociationRequestTimedOut`
This would allow consumers to count the number of times this happens and then set a CancellationTokenSource to canceled, which would cancel `SendAsync`. This is just one option.

What do you think @BobSter3000 ? I'm open to suggestions.

gofal commented:
Agree. Or a property MaxRetryCount, which is very common in many libraries.

amoerie commented:
I like the idea of a MaxRetryCount, but then we do need to think what should happen when a MaxRetryCount = 3 and a new request is added after attempt 2. Does the retry count reset or not? 
If the next attempt fails, is that then failure number 3 (for the already queued requests) or failure number 1 (for the newly added request)…

I think resetting the counter makes the most sense. Also, I believe that the number of people who add requests while already sending will be a minority.

BobSter3000 commented:
I lke the idea of the event and letting the consumer decide how to handle it with an event,  Just like the RequestTimed Event.  I will see what I can come up with.

BobSter3000 committed:
Implement Association Request retry count and event

BobSter3000 commented:
@amoerie First attempt at implemeting an event that fires when the Assoication Request times out.  There is also a max retry count for the association requests.   If the retry count exceeds the max retry count the exception is thrown.  I am not sure if this is OK or not.
Default Max Retry Count = 3.
Association Request sent
Association Request not responded within timeout Event Fired and Retry Count updated retry
Once Max Retry count is exceeded TransitionToCompletedWithErrorState with the timeout exception

amoerie commented on FO-DICOM.Core/Network/Client/DicomClientOptions.cs:
```diff
@@ -27,6 +27,11 @@ public class DicomClientOptions
         public int? MaximumNumberOfRequestsPerAssociation { get; set; } = null;
 
+        /// <summary>
+        /// The maximum number of Association Request Attempts after AssociationRequestTimeoutInMs expires 
```
I would rephrase this to "The maximum number of times an association request will be sent before giving up (in the event of a timeout, the request will be retried),  use <see cref="AssociationRequestTimeoutInMs" /> to configure the timeout"

amoerie commented on FO-DICOM.Core/Network/Client/States/DicomClientRequestAssociationState.cs:
```diff
@@ -204,7 +204,17 @@ public override async Task<IDicomClientState> GetNextStateAsync(DicomClientCance
+                _dicomClient.NotifyAssociationRequestTimedOut(new EventArguments.AssociationRequestTimedOutEventArgs(_dicomClient.ClientOptions.AssociationRequestTimeoutInMs, _dicomClient.AssociationRequestRetryCount));
+                if (_dicomClient.AssociationRequestRetryCount > _dicomClient.ClientOptions.MaximumNumberOfAssociationRequestAttempts)
+                {
+                    var exception = new DicomAssociationRequestTimedoutException();
```
I think you have to reset the AssociationRequestRetryCount here. Otherwise, in future SendAsync invocations, the request will not be retried.

amoerie replied:
You will also have to reset this count if the association request does NOT timeout (accept or reject)

amoerie commented on FO-DICOM.Core/Network/Client/EventArguments/AssociationRequestTimedOutEventArgs.cs:
```diff
@@ -0,0 +1,31 @@
+        /// <summary>
+        /// Gets the timeout that was exceded
+        /// </summary>
+        public int Timeout { get; }
```
Please either use a TimeSpan or rename the property to TimeoutInMs to make it explicit what the unit of time is.

amoerie reviewed (changes requested):
The code changes look good, but I think we should add two missing unit tests:

1. One new test that verifies this new exception is thrown after 3 timeouts. 
2. One new test that verifies association requests are always tried 3 times. 

If you would like some assistance with writing the tests, please let me know. I can give pointers or help you set something up. If you want to take a stab at it, you can find a lot of scenarios in the DicomClientTests and DicomClientTimeoutTests.

P.S. The code coverage check is failing, but it will succeed if you pull in the latest changes from development. We only just enabled code coverage, and the default settings do not allow ANY decrease in coverage. This is very harsh. I changed the config to allow a decrease of 0.1% (which is already more considerable than it sounds);

BobSter3000 commented:
@amoerie I will get started on these changes.  Thank you.

BobSter3000 committed:
Added parameters to the AssociationRequestTimedoutException
Reset counter in Associaiton Request State
Refactor ClientOptions

BobSter3000 committed:
Added Association Request Timout Unit Test to cover exception thrown and retry count

BobSter3000 committed:
Added Unit Test to cover exception thrown and retry count

BobSter3000 committed:
Merge branch 'fo-dicom:development' into AssociationReqTimeout

BobSter3000 committed:
Merge branch 'AssociationReqTimeout' of https://github.com/Bobster3000/fo-dicom into AssociationReqTimeout

BobSter3000 commented:
@amoerie I am having a problem trying to setup the second unit test requested.
2. One new test that verifies association requests are always tried 3 times.
I just do not see a way to see when the assocication is being requested again.  Maybe one of the state changed events?
The other unit test covers the exception and verifies the retry count.

amoerie commented:
I can help you out with that but please allow me some time. Thanks for the effort so far!

BobSter3000 commented:
@amoerie Thanks again for all the help on this.

BobSter3000 committed:
Fixed AssociationTimedOut Exception Message format

BobSter3000 committed:
Merge branch 'fo-dicom:development' into AssociationReqTimeout

BobSter3000 committed:
Merge branch 'AssociationReqTimeout' of https://github.com/Bobster3000/fo-dicom into AssociationReqTimeout

amoerie committed:
Merge development

amoerie committed:
Improve phrasing in DicomAssociationRequestTimedOutException

amoerie committed:
Improve docs in DicomClientOptions

amoerie committed:
Improve naming of association request timeout related properties

Also reset the counter at the appropriate times

amoerie committed:
Setup more configurable echo provider for association request case

amoerie committed:
Use Task.Run instead of Task.Factory.StartNew

StartNew actually doesn't support an asynchronous lambda and completes too soon. The consequences of this were minimal, but the logging is just confusing.

amoerie committed:
Use better stack traces in client completed state

amoerie committed:
Improve logging when association requests time out

amoerie committed:
Don't recursively enter the completed with error state

amoerie committed:
Add missing using

amoerie committed:
Add tests that verify the retry counter is reset when an association is accepted or rejected

amoerie commented:
@BobSter3000 I added two tests that verify the counter is reset when an association is accepted or rejected. These confirm that, when a timeout occurs, the same DicomClient can be reused again and will again try 2 or 3 times to associate.

You might notice that I also made some changes in seemingly unrelated parts. One is that the DicomClientConnectState was erroneously catching exception from the transition to the association request state, which since this PR starts throwing the timeout exception. This was wrong, and I moved the TransitionToAssociationRequestState outside the try block. The effect of the original code was that the completed with error state was entered multiple times. This caught my eye in the logs, and is now fixed.

Furthermore, and this was really subtle, I could see that the DicomClient connection cleanup would complete before the listener actually stops. I traced this back to the usage of TaskFactory.StartNew, which doesn't support await lambdas that return a Task. The result was that the Listener always completed much too quickly, causing confusing timings in the DicomClient logs. The solution was extremely simple: use Task.Run instead of TaskFactory.StartNew. We have lost the ability now to pass the flag "LongRunning", which hints at the TaskScheduler that a new dedicated thread should be launched, while Task.Run will borrow a thread from the threadpool. However, after reading up on some of the internal heuristics of the ThreadPool, it will also spin up extra threads if it sees some borrowed threads are long running, so ultimately the result will be the same as far as I can tell.

Sorry for "hijacking" your pull request to include these fixes.

amoerie committed:
Merge branch 'development' into AssociationReqTimeout

BobSter3000 commented:
@amoerie I really do appreciate all the.  Thanks again!

mrbean-bremen reviewed (approved):
LGTM

amoerie commented:
Merged! 🎉

Thanks for contributing @BobSter3000 !
